### PR TITLE
Console based queue dumps

### LINKED ISSUE #2546: Replace signal-based queue dumps with console based one
As soon as debug console support is merged, the USR1/USR2 way of dumping the event queue should be unified by replacing it with an additional command on the debug console.

### PR DESCRIPTION
**Pull request by marc-casperlabs**
Replaces USR1 dumps with console based ones. Closes #2546

### CODE REVIEW

File: node/src/components/console/tasks.rs [hunk 1]
```diff
@@ -1,24 +1,32 @@
 use std::{
     borrow::Cow,
     fmt::{self, Debug, Display, Formatter},
-    fs, io,
+    fs,
+    io::{self, Seek, SeekFrom},
     path::PathBuf,
 };
 
 use crate::{
```
Comment (Fraser999) [inline node/src/components/console/tasks.rs]: Can we sort out the import ordering pls?

File: node/src/effect/announcements.rs [hunk 2]
```diff
@@ -43,6 +47,26 @@ pub(crate) enum ControlAnnouncement {
...
     },
+    // An external event queue dump has been requested.
+    QueueDump {
+        /// One-shot serializer for queue dumps.
+        #[serde(skip)]
+        serializer: RefCell<Option<TempFileSerializer>>,
+    },
+}
+
+impl std::fmt::Debug for ControlAnnouncement {
```
Comment (Fraser999) [inline node/src/effect/announcements.rs]:
```suggestion
impl Debug for ControlAnnouncement {
```

**Review by Fraser999 — COMMENTED** [review]
This is looking good, but could you look to add a few tests please?  Seems like there's scope to add some unit tests as well as a couple of integration tests.

Also, renaming from "console" would be helpful.

File: node/src/components/console.rs [hunk 3]
```diff
@@ -17,17 +21,18 @@ use tracing::debug;
...
     types::NodeRng,
     utils::umask,
     WithDir,
 };
 use util::ShowUnixAddr;
 
-mod command;
-mod tasks;
-mod util;
+pub use tasks::TempFileSerializer;
```
Comment (rafal-ch) [inline node/src/components/console.rs]: That should possibly go between `use tracing::debug;` and `use super::Component;`

File: node/src/components/console/tasks.rs [hunk 4]
```diff
@@ -176,6 +225,73 @@ impl Session {
                             }
                         }
                     }
+                    Action::DumpQueues => {
+                        match tempfile::tempfile() {
+                            Ok(mut tmp) => {
+                                // We create a second handle to the serializer temporary file. It
+                                // will be discard once serialization is finished.
```
Comment (rafal-ch) [inline node/src/components/console/tasks.rs]:
```suggestion
                                // will be discarded once serialization is finished.
```

File: node/src/components/console/tasks.rs [hunk 5]
```diff
@@ -86,6 +94,29 @@ impl Display for Session {
     }
 }
 
+pub enum TempFileSerializer {
```
Comment (rafal-ch) [inline node/src/components/console/tasks.rs]: That looks like _maybe_ it could go into the `util.rs` of the `console` module?

**Review by rafal-ch — APPROVED** [review]
LGTM with just few minor nits :+1: 

I agree with @Fraser999 that at least some basic tests could come in handy.

File: node/src/components/console/tasks.rs (hunk 1 above)
Comment (marc-casperlabs) [reply node/src/components/console/tasks.rs]: I'm assuming you mean the ordering of blocks, which is supposed to be `std`, third party crates, crate local ones?

File: node/src/components/console/tasks.rs (hunk 5 above)
Comment (marc-casperlabs) [reply node/src/components/console/tasks.rs]: It's a little it too specific in my mind. `util` to me is things that, if they were more numerous, could be in a separate crate.

File: node/src/components/console/tasks.rs (hunk 1 above)
Comment (Fraser999) [reply node/src/components/console/tasks.rs]: Four sections: std, third party, other casper, local

File: node/src/utils/round_robin.rs [hunk 6]
```diff
@@ -115,70 +113,31 @@ struct Slot<K> {
...
 
 impl<I, K> WeightedRoundRobin<I, K>
 where
     I: Debug,
     K: Copy + Clone + Eq + Hash + IntoEnumIterator + Debug,
 {
-    /// Dump the contents of the queues (`Debug` representation) to a given file.
-    pub async fn debug_dump(&self, file: &mut File) -> Result<(), io::Error> {
+    /// Dump the queue contents to the given dumper function.
+    pub async fn dump<F: FnOnce(&QueueDump<K, I>)>(&self, dumper: F) {
```
Comment (goral09) [inline node/src/utils/round_robin.rs]: So that is a shorthand for a function that returns `Unit`? Didn't know that.

**Review by goral09 — DISMISSED** [review]
Seems a bit complicated but I guess that's inevitable. Nicely done, TIL :)

File: node/src/utils/round_robin.rs (hunk 6 above)
Comment (marc-casperlabs) [reply node/src/utils/round_robin.rs]: Yes, the `-> ()` is optional. I'd wager you're writing the shorthand every time you define a function without a return type already :)

**Comment by marc-casperlabs** [comment]
> Seems a bit complicated but I guess that's inevitable.

You will be pleased to know that I found a way to get rid of the RefCell :)

**Comment by goral09** [comment]
> > Seems a bit complicated but I guess that's inevitable.
> 
> You will be pleased to know that I found a way to get rid of the RefCell :)

And it seems like it did not require changing any other code besides simply removing `RefCell`?

**Comment by marc-casperlabs** [comment]
> And it seems like it did not require changing any other code besides simply removing RefCell?

I had to add a way to convert an event into a `ControlAnnouncement` owned, instead of just getting a by-ref conversion.

**Comment by marc-casperlabs** [comment]
Tests have been added!

**Comment by marc-casperlabs** [comment]
> Where are you verifying that the event to dump queues has been handled properly?

Fair enough, let me fix that.

File: node/src/effect/announcements.rs [hunk 7]
```diff
@@ -44,6 +48,49 @@ pub(crate) enum ControlAnnouncement {
         /// Error message.
         msg: String,
     },
+    // An external event queue dump has been requested.
+    QueueDumpRequest {
```
Comment (goral09) [inline node/src/effect/announcements.rs:52]: Why is the other request for diagonostic console not part of this `ControlAnnouncement` enum?

File: node/src/testing/network.rs [hunk 8]
```diff
@@ -28,9 +30,20 @@ use crate::{
...
-    fn node_id(&self) -> NodeId;
+    ///
+    /// The default implementation generates a pseudo-id base on its memory address.
+    fn node_id(&self) -> NodeId {
+        #[allow(trivial_casts)]
+        let addr = self as *const _ as usize;
+        let mut raw: [u8; KeyFingerprint::LENGTH] = [0; KeyFingerprint::LENGTH];
+        raw[0..(mem::size_of::<usize>())].copy_from_slice(&addr.to_be_bytes());
+        NodeId::from(KeyFingerprint::from(raw))
+    }
```
Comment (goral09) [inline node/src/testing/network.rs:46]: Just for my own understanding - how is this change related to the task? Did you stumble across this bit when testing the queue dump functionality?

File: node/src/components/diagnostics_port.rs [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        debug!(socket_path=%socket_path.display(), "found stale socket file, trying to remove");
+        match fs::remove_file(&socket_path) {
+            Ok(_) => {
+                debug!("stale socket file removed");
+            }
+            Err(err) => {
+                // This happens if a background tasks races us for the removal, as it usually
+                // means the file is already gone. We can ignore this, but make note of it in
+                // the log.
+                debug!(%err, "could not remove stale socket file, assuming race with background task");
```
Comment (goral09) [inline node/src/components/diagnostics_port.rs]: Should this be `warn` at least?

File: node/src/components/diagnostics_port.rs [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+            .accept()
+            .await
+            .expect("could not accept connection");
+
+        let mut buffer = Vec::new();
+        stream
+            .read_to_end(&mut buffer)
+            .await
+            .expect("failed to read to end");
+        assert_eq!(b"hello", buffer.as_slice());
```
Comment (goral09) [inline node/src/components/diagnostics_port.rs]: Nitpick but `hello` could be extracted to a variable and shared between the writing task and here.

File: node/src/components/diagnostics_port/tasks.rs [hunk 11]
```diff
@@ -0,0 +1,630 @@
...
+    fs::{self, File},
+    io,
+    path::PathBuf,
+};
+
+use bincode::{
+    config::{AllowTrailing, FixintEncoding, WithOtherIntEncoding, WithOtherTrailing},
+    DefaultOptions, Options,
+};
+use erased_serde::Serializer as ErasedSerializer;
```
Comment (goral09) [inline node/src/components/diagnostics_port/tasks.rs:13]: Smart!

File: node/src/components/diagnostics_port/tasks.rs [hunk 12]
```diff
@@ -0,0 +1,630 @@
...
+enum ObtainDumpError {
+    /// Error trying to create a temporary directory.
+    #[error("could not create temporary directory")]
+    CreateTempDir(#[source] io::Error),
+    /// Error trying to create a file in the temporary directory.
+    #[error("could not create file in temporary directory")]
+    CreateTempFile(#[source] io::Error),
+    /// Error trying to reopen the file in the temporary directory after writing.
+    #[error("could not reopen file in temporary directory")]
+    ReopenTempFile(#[source] io::Error),
```
Comment (goral09) [inline node/src/components/diagnostics_port/tasks.rs:138]: Where is this `#[source]` coming from? What crate? Will it be included in the error output?

File: node/src/components/diagnostics_port/tasks.rs [hunk 13]
```diff
@@ -0,0 +1,630 @@
...
+    ///
+    /// The resulting serializer will write to the given file.
+    fn create_queue_dump_format(&self, file: File) -> QueueDumpFormat {
+        match self.output {
+            OutputFormat::Interactive => QueueDumpFormat::debug(file),
+            OutputFormat::Json => {
+                QueueDumpFormat::serde(FileSerializer::Json(serde_json::Serializer::new(file)))
+            }
+            OutputFormat::Bincode => {
+                QueueDumpFormat::serde(FileSerializer::Bincode(bincode::Serializer::new(
```
Comment (goral09) [inline node/src/components/diagnostics_port/tasks.rs:178]: I wonder, whether we would ever want to use `bincode` - it's not very human readable.

File: node/src/components/diagnostics_port/tasks.rs [hunk 14]
```diff
@@ -0,0 +1,630 @@
...
+    async fn process_line<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        writer: &mut OwnedWriteHalf,
+        line: &str,
+    ) -> io::Result<bool>
+    where
+        REv: From<DumpConsensusStateRequest> + From<ControlAnnouncement> + Send,
+    {
+        debug!(%line, "line received");
```
Comment (goral09) [inline node/src/components/diagnostics_port/tasks.rs:200]: I know I already reviewed this code. Pity that I have to do it again. Instead, you could have made the changes first and then maybe do a rename as a separate PR.

**Review by goral09 — APPROVED** [review]

File: node/src/components/diagnostics_port.rs (hunk 9 above)
Comment (marc-casperlabs) [reply node/src/components/diagnostics_port.rs]: It could be external to the node as well. I'll upgrade the warning (and fix the typo in the comment as well :))

File: node/src/testing/network.rs (hunk 8 above)
Comment (marc-casperlabs) [reply node/src/testing/network.rs:46]: The test reactor for queue dumping has no networking component, but to use the `Network` testing structure it expects a `NetworkedReactor` implementation.

This is a bit of a misnomer, it does not need an actual network, but rather a unique ID per node. So instead of cramming a networking implementation into my tiny test-reactor, I instead decided to provide a unique ID based on something other than a random network ID, in this case the memory address of the instantiated node.

As it turns out, this makes a reasonable default implementation.

File: node/src/components/diagnostics_port/tasks.rs (hunk 11 above)
Comment (marc-casperlabs) [reply node/src/components/diagnostics_port/tasks.rs:13]: I learn from Fraser!

File: node/src/components/diagnostics_port/tasks.rs (hunk 12 above)
Comment (marc-casperlabs) [reply node/src/components/diagnostics_port/tasks.rs:138]: Its from `thiserror::Error`. It will generate a [`source()`](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) implementation (the default is to return `None`), which is just good practice.

There are basically two "styles" of chaining errors - one is to use the properly typed methods like `source()` (see link above), which will show up cleanly in an error trace, the other is just stuffing it into a string, e.g.

```rust
 #[error("could not reopen file in temporary directory: {0}")]
    ReopenTempFile(io::Error),
```

I prefer the former, the only drawback is that the default `Display` implementation generated by `thiserror` does not automatically add the `: {}`, as it may be used in an explicit backtrace-style printing (see the docs on `source()` above, contents of the main function).

I usually try to get the best of both worlds by using our helper function `display_error`.

File: node/src/components/diagnostics_port/tasks.rs (hunk 13 above)
Comment (marc-casperlabs) [reply node/src/components/diagnostics_port/tasks.rs:178]: For large dumps that we have special clients for.

We could drop it though, that's a good point; after all, we can on-the-fly gzip larger ones. :thinking:

File: node/src/components/diagnostics_port/tasks.rs (hunk 14 above)
Comment (marc-casperlabs) [reply node/src/components/diagnostics_port/tasks.rs:200]: I'm sorry :(

File: node/src/components/diagnostics_port/tasks.rs (hunk 12 above)
Comment (goral09) [reply node/src/components/diagnostics_port/tasks.rs:138]: I like this one more as well. Thanks.

File: node/src/components/diagnostics_port/tasks.rs (hunk 13 above)
Comment (goral09) [reply node/src/components/diagnostics_port/tasks.rs:178]: >For large dumps that we have special clients for.

We have clients for that? Who?

File: node/src/effect/announcements.rs (hunk 7 above)
Comment (marc-casperlabs) [reply node/src/effect/announcements.rs:52]: Because `ControlAnnouncement` (which should actually be named `ControlEvent` now, tbh) groups all events that are handled outside of the reactor by the runtime directly.

The other request is for data from consensus and requires no direct interventions from the `Runner`, as opposed to this one.

Comment (goral09) [reply node/src/effect/announcements.rs:52]: Ah, of course. Makes a lot of sense. Thanks.

File: node/src/components/diagnostics_port/tasks.rs (hunk 13 above)
Comment (marc-casperlabs) [reply node/src/components/diagnostics_port/tasks.rs:178]: Err, might write, sorry :)

**Comment by marc-casperlabs** [comment]
bors retry
